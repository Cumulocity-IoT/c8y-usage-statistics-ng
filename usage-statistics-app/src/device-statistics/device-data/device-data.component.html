<c8y-title>
  {{ "Device Data" | translate }}
</c8y-title>

<statistics-action-bar [feature]="feature"></statistics-action-bar>

<div class="content-fullpage d-flex d-col shadow1">
    <c8y-data-grid
    [loadMoreItemsLabel]="loadMoreItemsLabel"
    [loadingItemsLabel]="loadingItemsLabel"
    [columns]="columns"
    [rows]="deviceData"
    [pagination]="pagination"
    [infiniteScroll]="infiniteScroll"
    [actionControls]="actionControls"
    [showSearch]="showSearch"
    class="d-contents"
  >
    <div class="c8y-empty-state">
      <h1 c8yIcon="search"></h1>
      <div>
        <p>
          <strong>{{ "Device details are not available." | translate }}</strong>
        </p>
      </div>
    </div>
    <ng-container *ngFor="let column of columns">
      <c8y-column [name]="column.name"></c8y-column>
    </ng-container>
  </c8y-data-grid>
</div>
