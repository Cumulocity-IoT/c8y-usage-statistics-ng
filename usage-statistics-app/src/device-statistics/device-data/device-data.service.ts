import { Injectable } from "@angular/core";

export const DEVICE_ID: string = 'deviceId';
export const TOTAL_MEA: string = 'totalMea';
export const AVG_MEA: string = 'avgMea';
export const CLASS: string = 'className';
export const DEVICE_TYPE: string = 'deviceType';

@Injectable({
  providedIn: "root",
})
export class DeviceDataService {}
