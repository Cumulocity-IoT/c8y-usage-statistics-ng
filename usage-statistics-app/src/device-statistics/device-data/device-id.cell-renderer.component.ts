import { Component } from "@angular/core";
import { CellRendererContext } from "@c8y/ngx-components";

@Component({
  template: `
    <a
      href="/apps/devicemanagement/index.html#/device/{{ context.value }}"
      [title]="context.value"
      class="interact"
      target="_blank"
      rel="noopener noreferrer"
    >
      {{ context.value }}
    </a>
  `,
})
export class DeviceIdCellRendererComponent {
  constructor(
    public context: CellRendererContext,
  ) {}
}
