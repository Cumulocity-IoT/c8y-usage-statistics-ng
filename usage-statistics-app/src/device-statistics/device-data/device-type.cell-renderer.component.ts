import { Component } from "@angular/core";
import { CellRendererContext } from "@c8y/ngx-components";

@Component({
  template: `
    <span>{{ context.value }}</span>
  `,
})
export class DeviceTypeCellRendererComponent {
  constructor(
    public context: CellRendererContext,
  ) {}
}
