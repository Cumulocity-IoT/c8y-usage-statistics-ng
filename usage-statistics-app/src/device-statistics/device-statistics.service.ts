import { Injectable } from "@angular/core";
import { FetchClient } from "@c8y/client";
import { CommonService } from "../common.service";
import { DeviceConfigurationService } from "./device-configuration/device-configuration.service";
import { MonthPickerService } from "../utitities/statistics-action-bar/month-picker/month-picker.service";
const moment = require('moment')
const _ = require('lodash')
export const DATE_FORMAT_DAY = 'YYYY-MM-DD';

export const CLASS_COLORS = ['#FFBE00', '#119D11', '#00A1F2', '#FF8800', '#E51A1A', '#212121'];

export interface DeviceDataModel {
  deviceId: string,
  count: number,
  deviceType?: string
}
export interface FormattedDeviceDataModel extends DeviceDataModel {
  className: string,
  avgMea: string
}

@Injectable({
  providedIn: 'root'
})
export class DeviceStatisticsService {
  private formattedDeviceDataStore = {};
  private deviceClassDataStore = {};
  private deviceDataStore = {
    statistics: [],
    selectedDate: null
  };
  deviceStatisticsDataStore;

  constructor(
    private client: FetchClient,
    private commonService: CommonService,
    private deviceConfigurationService: DeviceConfigurationService,
    private monthPickerService: MonthPickerService
  ) { }

  async getFormattedDeviceData(selectedDate: Date) {

    const currentConfiguration = this.deviceConfigurationService.defaultConfig;
    const deviceData = await this.getDeviceData(selectedDate);
    
    
      this.formattedDeviceDataStore = {}
      this.deviceClassDataStore = {};
      currentConfiguration.forEach(config => this.deviceClassDataStore[config.className] = {
        threshold: config.monthlyThreshold || Number.POSITIVE_INFINITY,
        total: 0,
        className: config.className,
        deviceIdStore: []
      })

      deviceData.statistics.forEach(device => {
        this.populateFormattedDeviceDataStore(device)
        this.formattedDeviceDataStore[device.deviceId].avgMea =
          +((this.formattedDeviceDataStore[device.deviceId].totalMea / this.monthPickerService.daysInMonth).toFixed(2))
        const avgMea = this.formattedDeviceDataStore[device.deviceId].avgMea

        currentConfiguration.forEach((config, index) => {
          if ((avgMea >= config.avgMinMea && avgMea < config.avgMaxMea) ||
            (index === currentConfiguration.length - 1 && avgMea >= config.avgMinMea)) {
            this.formattedDeviceDataStore[device.deviceId]['className'] = config.className;
            this.deviceClassDataStore[config.className].total += device.count;
            this.deviceClassDataStore[config.className].deviceIdStore.push(device.deviceId);
          }
        })
      })

      this.deviceStatisticsDataStore = {
        deviceData: Object.values(this.formattedDeviceDataStore),
        overview: this.deviceClassDataStore,
        date: selectedDate
      }
    
    return this.deviceStatisticsDataStore
  }

  private populateFormattedDeviceDataStore(device) {
    if (this.formattedDeviceDataStore[device.deviceId]) {
      this.formattedDeviceDataStore[device.deviceId].totalMea += device.count;
    }
    else {
      this.formattedDeviceDataStore[device.deviceId] = {
        totalMea: device.count,
        deviceId: device.deviceId,
        deviceType: device.deviceType
      }
    }
  }


  private async getDeviceData(selectedDate: Date) {
    this.deviceDataStore = { statistics: [], selectedDate: null };
    let api = await this.getDeviceStatisticsUrl('monthly', moment(selectedDate).format(DATE_FORMAT_DAY));
    let statistics = undefined;
    let response = undefined;

    do {
      response = await this.getMonthlyDeviceStatistics(api, selectedDate)
      api = response.next ? response.next.substring(response.next.indexOf('/tenant')) : response.next;
      statistics = response.statistics;
      this.deviceDataStore.statistics.push(...response.statistics)
    }
    while (response.next && statistics && statistics.length)
    this.deviceDataStore.selectedDate = selectedDate;

    return this.deviceDataStore
  }

  private async getDeviceStatisticsUrl(period, duration) {
    return `tenant/statistics/device/${await this.commonService.getCurrentlyActiveTenant()}/${period}/${duration}?pageSize=2000`;
  }

  private async getMonthlyDeviceStatistics(api: string, date: Date) {
    try {
      const options = {
        method: 'GET'
      };
      const response = await this.client.fetch(api, options);
      return (await response).json();
    } catch (err) {
      console.log(err);
    }
  }

  public async getMonthlyDeviceAggregation(date: Date):Promise<any> {
    try {
      const options = {
        method: 'GET'
      };
      const response = await this.client.fetch(
        `/service/metrics-aggregator/devices/monthly/${moment(date).format(DATE_FORMAT_DAY)}?includeSubtenants=false`
        , options);
      return (await response).json();
    } catch (err) {
      console.log(err);
    }
  }


  public async getDailyDeviceAggregation(omitCache:boolean):Promise<any> {
    try {
      const options = {
        method: 'GET'
      };
      let url = "/service/metrics-aggregator/devices/dailystatistics";
      if (omitCache) {
        url = url + "?omitCache=true"
      }
      const response = await this.client.fetch(url, options);
      return (await response).json();
    } catch (err) {
      console.log(err);
    }
  }
}
