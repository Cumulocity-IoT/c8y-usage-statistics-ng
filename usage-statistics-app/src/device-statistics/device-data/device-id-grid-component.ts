import { Type } from '@angular/core';
import { Column, ColumnDataType, gettext } from '@c8y/ngx-components';
import { DeviceIdCellRendererComponent } from './device-id.cell-renderer.component';
import { DEVICE_ID } from './device-data.service';

export class DeviceIdGridColumn implements Column {
  name: string;
  path?: string;
  header?: string;
  dataType?: ColumnDataType;
  

  visible?: boolean;
  positionFixed?: boolean;
  gridTrackSize?: string;

  cellRendererComponent?: Type<any>;

  sortable?: boolean;

  filterable?: boolean;

  constructor() {
    this.name = DEVICE_ID;
    this.path = DEVICE_ID;
    this.header = gettext('Device ID');
    this.cellRendererComponent = DeviceIdCellRendererComponent;

    this.filterable = true;
    this.sortable = true;
  }
}
