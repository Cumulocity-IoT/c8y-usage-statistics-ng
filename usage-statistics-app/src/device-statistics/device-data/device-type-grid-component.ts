import { Type } from '@angular/core';
import { Column, ColumnDataType, gettext } from '@c8y/ngx-components';
import { DEVICE_TYPE } from './device-data.service';
import { DeviceTypeCellRendererComponent } from './device-type.cell-renderer.component';

export class DeviceTypeGridColumn implements Column {
  name: string;
  path?: string;
  header?: string;
  dataType?: ColumnDataType;

  visible?: boolean;
  positionFixed?: boolean;
  gridTrackSize?: string;

  cellRendererComponent?: Type<any>;

  sortable?: boolean;

  filterable?: boolean;

  constructor() {
    this.name = DEVICE_TYPE;
    this.path = DEVICE_TYPE;
    this.header = gettext('Device type');
    this.cellRendererComponent = DeviceTypeCellRendererComponent;

    this.filterable = true;
    this.sortable = true;
  }
}
