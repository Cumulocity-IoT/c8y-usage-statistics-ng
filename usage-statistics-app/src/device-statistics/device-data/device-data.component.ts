import { Component, ElementRef, OnDestroy, OnInit } from '@angular/core';
import { Column, DisplayOptions, gettext, Pagination } from '@c8y/ngx-components';
import { Subscription } from 'rxjs';
import { MonthPickerService } from '../../utitities/statistics-action-bar/month-picker/month-picker.service';
import { DeviceStatisticsService } from '../device-statistics.service';
import { DeviceIdGridColumn } from './device-id-grid-component';
import { TOTAL_MEA, AVG_MEA, CLASS } from './device-data.service';
import { DeviceTypeGridColumn } from './device-type-grid-component';
import { CommonService, FeatureList } from '../../src/../common.service';

@Component({
  selector: 'device-data',
  templateUrl: './device-data.component.html',
  styleUrls: ['./device-data.component.css']
})
export class DeviceDataComponent implements OnInit, OnDestroy {
  feature = FeatureList.DeviceStatistics
  monthChangedSubscription: Subscription;
  tenantChangedSubscription: Subscription;
  deviceData: any[];
  loadMoreItemsLabel: string = gettext('Load more device details');
  loadingItemsLabel: string = gettext('Loading device details...');
  displayOptions: DisplayOptions = {
    bordered: false,
    striped: true,
    filter: true,
    gridHeader: true,
    hover: true
  };

  columns: Column[] = this.getColumns();
  pagination: Pagination = this.getPagination();

  constructor(
    private commonService: CommonService,
    private monthPickerService: MonthPickerService,
    private deviceStatisticsService: DeviceStatisticsService,
    private element: ElementRef
  ) {
  }

  ngOnInit(): void {    
    this.monthChangedSubscription = this.monthPickerService.dateChanged.subscribe((selectedDate: Date) => {
      this.getDeviceData(selectedDate)
    })
    this.tenantChangedSubscription = this.commonService.tenantChanged.subscribe((tenantId: string) =>{
      this.getDeviceData()
    })
    this.getDeviceData();
  }

  ngAfterViewInit() {
    const btns = this.element.nativeElement.querySelectorAll('c8y-data-grid .btnbar-btn');
    btns.forEach((btn: HTMLElement) => {
      if (btn.querySelector('i[c8yIcon=refresh]')) btn.style.display = 'none';
    });
  }

  getColumns(): Column[] {
    return [
      new DeviceIdGridColumn(),
      new DeviceTypeGridColumn(),
      {
        name: TOTAL_MEA,
        header: gettext('Total MEAs'),
        path: TOTAL_MEA,
        filterable: true,
        sortable: true
      },
      {
        name: AVG_MEA,
        header: gettext('Average MEAs/day'),
        path: AVG_MEA,
        filterable: true,
        sortable: true
      },
      {
        name: CLASS,
        header: gettext('Class'),
        path: CLASS,
        filterable: true,
        sortable: true
      }
    ];
  }

  getPagination(): Pagination {
    return {
      pageSize: 10,
      currentPage: 1
    };
  }

  async getDeviceData(selectedDate = this.monthPickerService.selectedDate) {
    const data = await this.deviceStatisticsService.getFormattedDeviceData(selectedDate);
    this.deviceData = data.deviceData;
  }
 
  ngOnDestroy(): void {
    if(this.monthChangedSubscription){
      this.monthChangedSubscription.unsubscribe();
    }
    if(this.tenantChangedSubscription){
      this.tenantChangedSubscription.unsubscribe();
    }
  }
}
