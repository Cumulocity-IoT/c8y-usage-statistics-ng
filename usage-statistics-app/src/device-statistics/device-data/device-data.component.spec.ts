import { NO_ERRORS_SCHEMA } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { DeviceDataComponent } from './device-data.component';
import { DeviceStatisticsService } from '../device-statistics.service';
import { RouterTestingModule } from '@angular/router/testing';
import { CoreModule } from '@c8y/ngx-components';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';

describe('Test DeviceDataComponent', () => {
  let component: DeviceDataComponent;
  let fixture: ComponentFixture<DeviceDataComponent>;
  let deviceStatisticsService: DeviceStatisticsService;

  const deviceStatisticsServiceMock = {
    getFormattedDeviceData: jest.fn(),
}

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [RouterTestingModule, CoreModule.forRoot(), BrowserAnimationsModule],
      declarations: [DeviceDataComponent],
      schemas: [NO_ERRORS_SCHEMA],
      providers: [
        { provide: DeviceStatisticsService, useValue: deviceStatisticsServiceMock },
    ]
    });
    fixture = TestBed.createComponent(DeviceDataComponent);
    component = fixture.componentInstance;
    deviceStatisticsService = TestBed.inject(DeviceStatisticsService);
  });

  test('should be defined', () => {
    expect(fixture).toBeDefined();
  });

  test('should load expected component', () => {
    expect(fixture).toMatchSnapshot();
  });

  test('should call getFormattedDeviceData from service', async () => {
    fixture.detectChanges();
    const getFormattedDeviceDataSpy = jest.spyOn(deviceStatisticsService, 'getFormattedDeviceData');
    expect(getFormattedDeviceDataSpy).toHaveBeenCalled();
  });

  test('should give correct device data response', async() => {
    fixture.detectChanges();
    let today = new Date("01 Jan 2023");
    const selectedDate = new Date(today.setMonth(today.getMonth() - 1));
    const mockResponse = {
      'date': selectedDate,
      'deviceData': [
          {
              "totalMea": 1071056,
              "deviceId": "62948",
              "deviceType": "c8y_MQTTDevice",
              "avgMea": 34550.19,
              "className": "Class E"
          },
          {
              "totalMea": 1070958,
              "deviceId": "1118134",
              "deviceType": "c8y_MQTTDevice",
              "avgMea": 34547.03,
              "className": "Class E"
          },
          {
              "totalMea": 1070958,
              "deviceId": "1118135",
              "deviceType": "c8y_MQTTDevice",
              "avgMea": 34547.03,
              "className": "Class E"
          },
          {
          "totalMea": 1070958,
              "deviceId": "1118136",
              "deviceType": "c8y_MQTTDevice",
              "avgMea": 34547.03,
              "className": "Class E"
          },
          {
              "totalMea": 1070956,
              "deviceId": "1118138",
              "deviceType": "c8y_MQTTDevice",
              "avgMea": 34546.97,
              "className": "Class E"
          }
      ],
      'overview': {
          "Class A": {
              "threshold": null,
              "total": 0,
              "className": "Class A",
              "deviceIdStore": []
          },
          "Class B": {
              "threshold": null,
              "total": 0,
              "className": "Class B",
              "deviceIdStore": []
          },
          "Class C": {
              "threshold": null,
              "total": 0,
              "className": "Class C",
              "deviceIdStore": []
          },
          "Class D": {
              "threshold": null,
              "total": 0,
              "className": "Class D",
              "deviceIdStore": []
          },
          "Class E": {
              "threshold": null,
              "total": 3748449,
              "className": "Class E",
              "deviceIdStore": [
                  "62948",
                  "1118136",
                  "1118135",
              ]
          },
          "Class F": {
              "threshold": null,
              "total": 0,
              "className": "Class F",
              "deviceIdStore": []
          }
      }
    };
    jest.spyOn(deviceStatisticsService, 'getFormattedDeviceData').mockImplementation(()=>Promise.resolve(mockResponse))
    await component.getDeviceData(selectedDate);
    expect(component.deviceData).toEqual(mockResponse.deviceData);
  });
});