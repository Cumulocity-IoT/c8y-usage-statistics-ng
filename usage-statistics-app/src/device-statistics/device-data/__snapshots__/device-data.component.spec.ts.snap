// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test DeviceDataComponent should load expected component 1`] = `
<device-data
  columns={[Function Array]}
  commonService={[Function CommonService]}
  deviceStatisticsService={[Function Object]}
  displayOptions={[Function Object]}
  element={[Function ElementRef]}
  feature={[Function String]}
  loadMoreItemsLabel={[Function String]}
  loadingItemsLabel={[Function String]}
  monthPickerService={[Function MonthPickerService]}
  pagination={[Function Object]}
>
  <c8y-title>
  </c8y-title><statistics-action-bar /><div
    class="content-fullpage d-flex d-col shadow1"
  >
    <c8y-data-grid
      class="d-contents"
    >
      <div
        class="table-data-grid-scroll"
        data-cy="c8y-data-grid--table-data-grid-scroll"
      >
        <table
          cdk-table=""
          class="cdk-table table table-hover table-filtered-sorted table-data-grid large-padding"
          data-cy="c8y-data-grid--table"
          role="table"
        >
        </table>
      </div>
    </c8y-data-grid>
  </div>
</device-data>
`;
